src/brush.jsx

```diff
@@ -9,75 +9,371 @@
+                } else {
+                    newBegin = tb < viewport.begin().getTime() ? viewport.begin() : tb;
+                    newEnd = t > viewport.end().getTime() ? viewport.end() : t;
+                }
             } else {
-                return;
+                const xy0 = this.state.initialBrushXYPosition;
+                let timeOffset =
+                    this.props.timeScale.invert(xy0[0]).getTime() -
+                    this.props.timeScale.invert(xy[0]).getTime();
+
+                // Constrain
+                if (tb - timeOffset < viewport.begin()) {
+                    timeOffset = tb - viewport.begin().getTime();
+                }
+                if (te - timeOffset > viewport.end()) {
+                    timeOffset = te - viewport.end().getTime();
+                }
+
+                newBegin =
+                    this.state.brushingInitializationSite === "brush" ||
+                    this.state.brushingInitializationSite === "handle-left" ?
+                        parseInt(tb - timeOffset, 10) : tb;
+                newEnd =
+                    this.state.brushingInitializationSite === "brush" ||
+                    this.state.brushingInitializationSite === "handle-right" ?
+                        parseInt(te - timeOffset, 10) : te;
+
+                // Swap if needed
+                if (newBegin > newEnd) [newBegin, newEnd] = [newEnd, newBegin];
+            }
+
+            if (this.props.onTimeRangeChanged) {
+                this.props.onTimeRangeChanged(new TimeRange(newBegin, newEnd));
             }
         }
-        d3.select(ReactDOM.findDOMNode(this)).selectAll("*").remove();
-
-        d3.select(ReactDOM.findDOMNode(this))
-            .append("g")
-                .attr("class", "x brush")
-                .call(this.brush)
-                .selectAll("rect")
-                .attr("y", -6)
-                .attr("height", this.props.height + 7);
     },
 
-    componentDidMount() {
-        this.renderBrush(this.props.timeScale, this.props.timeRange);
+    renderOverlay() {
+        const { width, height } = this.props;
+
+        let cursor;
+        switch (this.state.brushingInitializationSite) {
+            case "handle-right":
+            case "handle-left":
+                cursor = "ew-resize";
+                break;
+            case "brush":
+                cursor = "move";
+                break;
+            default:
+                cursor = "crosshair";
+        }
+
+        const overlayStyle = {
+            fill: "white",
+            opacity: 0,
+            cursor
+        };
+        return (
+            <rect
+                ref="overlay"
+                x={0} y={0}
+                width={width} height={height}
+                style={overlayStyle}
+                onMouseDown={this.handleOverlayMouseDown}
+                onMouseUp={this.handleMouseUp}
+                onClick={this.handleClick}
+                clipPath={this.props.clipPathURL} />
```

Comment:
I guess what I mean is, is this a prop or state? I'll read a bit more, this may be a non-issue.